File: android/app/src/main/kotlin/com/example/amber_list/widget/WidgetSettingsHelper.kt
package com.example.amber_list.widget

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Widget Settings Helper
 *
 * Reads and writes task management settings from Flutter's SharedPreferences.
 * This allows the Widget to share the same auto-postpone state as the Flutter app.
 *
 * SharedPreferences file: FlutterSharedPreferences
 * Key: flutter.task_management_settings
 * Value: JSON object with fields:
 *   - enableAutoPostpone: boolean (default true)
 *   - overdueExpanded: boolean (default true)
 *   - lastAutoPostponeDate: string "yyyy-MM-dd HH:mm:ss" (nullable)
 */
object WidgetSettingsHelper {

    private const val TAG = "WidgetSettingsHelper"
    private const val PREFS_NAME = "FlutterSharedPreferences"
    private const val KEY = "flutter.task_management_settings"

    /**
     * Check if auto-postpone has been performed today
     *
     * Compares the date part (yyyy-MM-dd) of lastAutoPostponeDate with today's date.
     * Returns true if they match, meaning auto-postpone was already done today.
     *
     * @param context Application context
     * @return true if already checked today, false otherwise
     */
    fun hasCheckedToday(context: Context): Boolean {
        try {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val json = prefs.getString(KEY, null)

            if (json.isNullOrEmpty()) return false

            val obj = JSONObject(json)
            val lastDate = obj.optString("lastAutoPostponeDate", "")
            if (lastDate.isEmpty()) return false

            // Compare date part only (first 10 chars: yyyy-MM-dd)
            val today = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())
            val lastDatePart = if (lastDate.length >= 10) lastDate.substring(0, 10) else lastDate
            return lastDatePart == today
        } catch (e: Exception) {
            Log.e(TAG, "hasCheckedToday error", e)
            return false
        }
    }

    /**
     * Mark today as checked (auto-postpone has been performed)
     *
     * Updates lastAutoPostponeDate to current datetime in "yyyy-MM-dd HH:mm:ss" format.
     * Preserves other settings in the JSON.
     *
     * @param context Application context
     */
    fun setCheckedToday(context: Context) {
        try {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val existingJson = prefs.getString(KEY, null)

            // Parse existing or create new JSON object
            val obj = if (!existingJson.isNullOrEmpty()) {
                JSONObject(existingJson)
            } else {
                JSONObject()
            }

            // Update lastAutoPostponeDate
            val now = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(Date())
            obj.put("lastAutoPostponeDate", now)

            // Save back
            prefs.edit().putString(KEY, obj.toString()).apply()
        } catch (e: Exception) {
            Log.e(TAG, "setCheckedToday error", e)
        }
    }

    /**
     * Check if auto-postpone feature is enabled globally
     *
     * Reads enableAutoPostpone from settings. Defaults to true if not set.
     *
     * @param context Application context
     * @return true if auto-postpone is enabled, false otherwise
     */
    fun isAutoPostponeEnabled(context: Context): Boolean {
        try {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val json = prefs.getString(KEY, null)

            if (json.isNullOrEmpty()) return true // Default enabled

            val obj = JSONObject(json)
            return obj.optBoolean("enableAutoPostpone", true)
        } catch (e: Exception) {
            Log.e(TAG, "isAutoPostponeEnabled error", e)
            return true // Default enabled on error
        }
    }
}

File: ios/AmberWidget/Models/WidgetSettingsHelper.swift
//
//  WidgetSettingsHelper.swift
//  AmberWidget
//
//  Widget Settings Helper
//
//  Reads and writes task management settings from Flutter's shared storage.
//  This allows the Widget to share the same auto-postpone state as the Flutter app.
//
//  Storage location: App Group UserDefaults
//  Key: task_management_settings
//  Value: JSON object with fields:
//    - enableAutoPostpone: boolean (default true)
//    - overdueExpanded: boolean (default true)
//    - lastAutoPostponeDate: string "yyyy-MM-dd HH:mm:ss" (nullable)
//

import Foundation

/// Widget Settings Helper
/// Manages auto-postpone settings shared between Flutter App and iOS Widget
class WidgetSettingsHelper {

    // MARK: - Constants

    /// App Group ID (must match Flutter HomeWidgetService and Xcode config)
    private static let appGroupID = "group.com.amberlist.amberList"

    /// Settings key in UserDefaults
    private static let settingsKey = "task_management_settings"

    // MARK: - UserDefaults Access

    /// Get shared UserDefaults for App Group
    private static var sharedDefaults: UserDefaults? {
        UserDefaults(suiteName: appGroupID)
    }

    /// Load settings JSON from UserDefaults
    private static func loadSettingsJson() -> [String: Any]? {
        guard let defaults = sharedDefaults,
              let jsonString = defaults.string(forKey: settingsKey),
              !jsonString.isEmpty else {
            return nil
        }

        guard let jsonData = jsonString.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else {
            print("[WidgetSettings] Failed to parse settings JSON")
            return nil
        }

        return json
    }

    /// Save settings JSON to UserDefaults
    private static func saveSettingsJson(_ json: [String: Any]) {
        guard let defaults = sharedDefaults else {
            print("[WidgetSettings] Cannot access shared UserDefaults")
            return
        }

        guard let jsonData = try? JSONSerialization.data(withJSONObject: json),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            print("[WidgetSettings] Failed to serialize settings JSON")
            return
        }

        defaults.set(jsonString, forKey: settingsKey)
    }

    // MARK: - Auto-Postpone Settings

    /// Check if auto-postpone has been performed today
    ///
    /// Compares the date part (yyyy-MM-dd) of lastAutoPostponeDate with today's date.
    /// Returns true if they match, meaning auto-postpone was already done today.
    ///
    /// - Returns: true if already checked today, false otherwise
    static func hasCheckedToday() -> Bool {
        guard let json = loadSettingsJson(),
              let lastDate = json["lastAutoPostponeDate"] as? String,
              !lastDate.isEmpty else {
            print("[WidgetSettings] hasCheckedToday: no lastAutoPostponeDate found")
            return false
        }

        // Compare date part only (first 10 chars: yyyy-MM-dd)
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let today = formatter.string(from: Date())

        let lastDatePart = lastDate.count >= 10 ? String(lastDate.prefix(10)) : lastDate
        let checked = lastDatePart == today

        print("[WidgetSettings] hasCheckedToday: \(checked) (lastDate=\(lastDatePart), today=\(today))")
        return checked
    }

    /// Mark today as checked (auto-postpone has been performed)
    ///
    /// Updates lastAutoPostponeDate to current datetime in "yyyy-MM-dd HH:mm:ss" format.
    /// Preserves other settings in the JSON.
    static func setCheckedToday() {
        var json = loadSettingsJson() ?? [:]

        // Update lastAutoPostponeDate
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        let now = formatter.string(from: Date())

        json["lastAutoPostponeDate"] = now

        saveSettingsJson(json)
        print("[WidgetSettings] setCheckedToday: \(now)")
    }

    /// Check if auto-postpone feature is enabled globally
    ///
    /// Reads enableAutoPostpone from settings. Defaults to true if not set.
    ///
    /// - Returns: true if auto-postpone is enabled, false otherwise
    static func isAutoPostponeEnabled() -> Bool {
        guard let json = loadSettingsJson() else {
            // Default enabled if no settings found
            return true
        }

        // Default to true if key not present
        let enabled = json["enableAutoPostpone"] as? Bool ?? true
        print("[WidgetSettings] isAutoPostponeEnabled: \(enabled)")
        return enabled
    }

    // MARK: - Debug

    /// Get last auto-postpone check date (for debugging)
    ///
    /// - Returns: lastAutoPostponeDate string or nil
    static func getLastCheckDate() -> String? {
        guard let json = loadSettingsJson(),
              let lastDate = json["lastAutoPostponeDate"] as? String else {
            return nil
        }
        return lastDate
    }
}
